platforms/core-configuration/model-core/src/main/java/org/gradle/model/internal/manage/schema/extract/ScalarCollectionNodeInitializerExtractionStrategy.java

```diff
@@ -342,15 +338,34 @@ public List<T> subList(int fromIndex, int toIndex) {
         @Override
         public T set(int index, T element) {
             validateElementType(element);
-            return ((List<T>)getDelegate(true)).set(index, element);
+            return getDelegate(true).set(index, element);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            return getDelegate(false).equals(o);
```

Comment:
💭 ErrorProne complains about `Collection.equals` because it has a very weak contract. I'm following its recommendation and replace `Collection.equals` with explicit `List.equals` and `Set.equals` here and below, in `SetBackedCollection`. At runtime it still works the same, of course.